### ISSUE TITLE
Dataset Class

### ISSUE THREAD

**Issue by ZahraDalvand** [suprosanna/relationformer#24]
**class FloorplanGraphDataset(Dataset):
    """**
Dataset class for RelationFormer
    ./dataset/{train, val, test}/
        ├── images/scene_00000_to_image_rational.png
        └── masks/annotation_2d_scene_00000.json
    """
    def __init__(self, root_path="./dataset", split="train"):
        assert split in ["train", "val", "test"], "split باید یکی از train/val/test باشد"

        self.images_dir = os.path.join(root_path, split, "images")
        self.masks_dir = os.path.join(root_path, split, "masks")

        # Only files that have both an image and a corresponding mask are included.
        self.ids = sorted([
            f.replace("scene_", "").replace("_to_image_rational.png", "")
            for f in os.listdir(self.images_dir)
            if f.endswith("_to_image_rational.png")
            and os.path.exists(
                os.path.join(
                    self.masks_dir,
                    f"annotation_2d_scene_{f.replace('scene_', '').replace('_to_image_rational.png','')}.json"
                )
            )
        ])

        self.images = []
        self.nodes = []
        self.edges = []
        self._load_data()

    def _load_data(self):
        """
       PNG images and corresponding graph annotations are loaded together.
        """
        to_tensor = T.ToTensor()   

        for id_ in self.ids:
            # ----------- images -----------
            img_path = os.path.join(self.images_dir, f"scene_{id_}_to_image_rational.png")
            img_pil = Image.open(img_path).convert("RGB")  # فقط 3 کانال
            img = to_tensor(img_pil)  # (3,H,W)
            img = torch.flip(img, dims=[1])  # معادل flipud برای محور H
            self.images.append(img)

            # ----------- graph mask -----------
            mask_path = os.path.join(self.masks_dir, f"annotation_2d_scene_{id_}.json")
            with open(mask_path, "r") as f:
                data = json.load(f)

            # --- Nodes are extracted from the annotations and sorted according to their IDs. ---
            junctions = data["junctions"]
            coords = [[j["coordinate"][0], j["coordinate"][1]] for j in junctions]
            coords = torch.FloatTensor(coords)

            # The nodes are ordered according to their IDs.
            node_ids = [j["ID"] for j in junctions]  
            node_ids_tensor = torch.tensor(node_ids)
            sort_idx = torch.argsort(node_ids_tensor)
            coords = coords[sort_idx]
            self.nodes.append(coords)

            # ---Edges are extracted from the incidence matrix, using the node ordering defined by their IDs. ---
            mat = np.array(data["lineJunctionMatrix"])
            edge_list = []
            for r in mat:
                nodes_idx = np.where(r == 1)[0]
                if len(nodes_idx) == 2:
                    # Indices are adjusted according to `sort_idx` to maintain consistent ordering of nodes and edges.
                    new_idx0 = (sort_idx == nodes_idx[0]).nonzero(as_tuple=True)[0].item()
                    new_idx1 = (sort_idx == nodes_idx[1]).nonzero(as_tuple=True)[0].item()
                    edge_list.append([new_idx0, new_idx1])
            edges_tensor = torch.tensor(edge_list, dtype=torch.long)
            self.edges.append(edges_tensor)

    def __len__(self):
        return len(self.ids)

    def __getitem__(self, idx):
        """
        خروجی مشابه RelationFormer + نام تصویر
        return:
          img[None], nodes, edges, id_, file_name
        """
        img = self.images[idx][None]   # شکل (1,3,H,W)
        nodes = self.nodes[idx]
        edges = self.edges[idx]
        id_ = self.ids[idx]
        file_name = f"scene_{id_}"
        return img, nodes, edges, id_, file_name

from torch.nn.utils.rnn import pad_sequence

**def custom_collate_fn(batch):**
    """
    Input  batch:
      img, nodes, edges, id_, file_name
    """
    imgs, nodes_list, edges_list, ids_list, file_names = zip(*batch)

    # پد کردن nodes و edges
    nodes_padded = pad_sequence(nodes_list, batch_first=True, padding_value=0.0)
 
    edges_padded = edges_list

    imgs = torch.stack(imgs)

    return imgs, nodes_padded, edges_padded, ids_list, file_names
